syne_tune/optimizer/schedulers/searchers/random_grid_searcher.py

```diff
@@ -329,22 +329,21 @@ def _next_candidate_on_grid(self) -> Optional[dict]:
         """
 
         num_combinations = len(self.hp_values_combinations)
-        if self._next_index < num_combinations:
-            is_not_done = True
-            while is_not_done and self._next_index < num_combinations:
-                candidate = dict(
-                    zip(self.hp_keys, self.hp_values_combinations[self._next_index])
-                )
-                self._next_index += 1
-                is_not_done = self._all_initial_configs.contains(candidate)
-            if self._allow_duplicates and self._next_index == num_combinations:
-                # Another round through the grid
-                self._next_index = 0
-                self._all_initial_configs = ExclusionList.empty_list(self._hp_ranges)
-            return candidate
-        else:
-            # No more candidates
-            return None
+        candidate = None
+        is_not_done = True
+        while is_not_done and self._next_index < num_combinations:
+            candidate = dict(
+                zip(self.hp_keys, self.hp_values_combinations[self._next_index])
+            )
+            self._next_index += 1
+            is_not_done = self._all_initial_configs.contains(candidate)
+        if self._allow_duplicates and self._next_index == num_combinations:
+            # Another round through the grid
+            self._next_index = 0
+            self._all_initial_configs = ExclusionList.empty_list(self._hp_ranges)
```

Comment:
I know this is pre-existing logic, but is it correct that the `_all_initial_configs` variable is being changed? Seems counterintuitive to me that this would be modified. Wouldn't it make more sense to leave `_all_initial_configs` as-is, and then over the course of the evaluation, modify some other variable such as `_remaining_configurations`?

Reply:
This is a bit subtle. In general, we enter ``points_to_evaluate`` into ``_all_initial_configs`` as they get suggested. Once this is done, we don't want to suggest these configs again in the first round over the grid.

If we allow for duplicates, there can be a second round over the grid, and there we can suggest the initial configs again (in the order they appear in the grid). This is why I need to reset ``_all_initial_configs`` here.